src/script/src/python/engine.rs

```diff
@@ -36,11 +43,117 @@ use crate::python::error::{self, Result};
+        engine.register_function(zelf)
+    }
+
+    /// Fake a schema, should only be used with dynamically eval a Python Udf
+    fn fake_schema(&self, columns: &[VectorRef]) -> SchemaRef {
+        let arg_names = &self.copr.deco_args.arg_names;
+        let col_sch: Vec<_> = columns
+            .iter()
+            .enumerate()
+            .map(|(i, col)| ColumnSchema::new(arg_names[i].to_owned(), col.data_type(), true))
+            .collect();
+        let schema = datatypes::schema::Schema::new(col_sch);
+        Arc::new(schema)
+    }
+}
+
+impl Function for PyUdf {
+    fn name(&self) -> &str {
+        &self.copr.name
+    }
+
+    fn return_type(
+        &self,
+        _input_types: &[datatypes::prelude::ConcreteDataType],
+    ) -> common_query::error::Result<datatypes::prelude::ConcreteDataType> {
+        // TODO: use correct return annotation if exist
+        Ok(self.copr.return_types[0]
+            .clone()
+            .map_or(ConcreteDataType::float64_datatype(), |anno| {
+                anno.datatype
+                    .unwrap_or(ConcreteDataType::float64_datatype())
+            }))
+    }
+
+    fn signature(&self) -> common_query::prelude::Signature {
+        Signature::uniform(
+            self.copr.arg_types.len(),
+            ConcreteDataType::numerics(),
+            Volatility::Immutable,
+        )
+    }
+
+    fn eval(
+        &self,
+        _func_ctx: common_function::scalars::function::FunctionContext,
+        columns: &[datatypes::vectors::VectorRef],
+    ) -> common_query::error::Result<datatypes::vectors::VectorRef> {
+        // FIXME: exec_parsed require a RecordBatch(basically a Vector+Schema), where schema can't pop out from nowhere, right?
+        let schema = self.fake_schema(columns);
+        let columns = columns.to_vec();
+        // TODO(discord9): remove unwrap
+        let rb = RecordBatch::new(schema, columns).context(UdfTempRecordBatchSnafu)?;
+        let res = exec_parsed(&self.copr, &rb).map_err(|err| {
+            PyUdfSnafu {
+                msg: format!("{err:#?}"),
+            }
+            .build()
+        })?;
+        let len = res.columns().len();
+        if len != 1 {
+            logging::info!("Python UDF should return exactly one column, found {len} column(s)");
+            if len == 0 {
+                return PyUdfSnafu {
+                    msg: format!(
+                        "Python UDF should return exactly one column, found {len} column(s)"
+                    ),
+                }
+                .fail();
+            } // if more than one columns, just return first one
+        }
+        // TODO(discord9): more error handling
+        let res0 = res.column(0);
+        Ok(res0.to_owned())
+    }
+}
+
 pub struct PyScript {
     query_engine: QueryEngineRef,
     copr: CoprocessorRef,
 }
 
+impl PyScript {
+    /// Register Current Script as UDF, register name is same as script name
+    /// FIXME(discord9): possible inject attack?
```

Comment:
maybe we should consider eval python script in some kind of sandbox in the future.